File: erlang/calcul.erl
-module(calcul).
-export([calcul/1]).

calcul(N) ->
    "n = " ++ erlang:integer_to_list(N) ++
    "\tReclangles\tTrapezes\tSimpson\n" ++
    "f1\t" ++ calcul_integ(fun(X) -> f1(X) end, N) ++
    "f2\t" ++ calcul_integ(fun(X) -> f2(X) end, N) ++
    "f3\t" ++ calcul_integ(fun(X) -> f3(X) end, N) ++
    "f4\t" ++ calcul_integ(fun(X) -> f4(X) end, N) ++
    "f5\t" ++ calcul_integ(fun(X) -> f5(X) end, N) ++
    "f6\t" ++ calcul_integ(fun(X) -> f6(X) end, N) ++
    "f7\t" ++ calcul_integ(fun(X) -> f7(X) end, N) ++
    "f8\t" ++ calcul_integ(fun(X) -> f8(X) end, N) ++
    "f9\t" ++ calcul_integ(fun(X) -> f9(X) end, N) ++
    "f10\t" ++ calcul_integ(fun(X) -> f10(X) end, N).

calcul_integ(F, N) ->
    A = 0.0,
    B = 1.0,
    H = (B - A) / N,

    io_lib:format("~.2e", [rectangles(F, N, A, H)]) ++ "\t\t" ++
    io_lib:format("~.2e", [trapezes(F, N, A, B, H)]) ++ "\t\t" ++
    io_lib:format("~.2e", [simpson (F, N, A, B, H)]) ++ "\n".

somme(I, E, F) ->
    if
        I < E -> F(I) + somme(I + 1, E, F);
        true -> 0.0
    end.

rectangles(F, N, A, H) ->
    Inner_somme = fun(I) -> F(A + H * I) end,
    S = somme(0, N, Inner_somme),

    abs((H * S) - 1.0).

trapezes(F, N, A, B, H) ->
    Inner_somme = fun(I) -> F(A + H * I) end,
    S = somme(1, N, Inner_somme),

    abs((((B - A) / (2.0 * N)) * (F(A) + F(B) + (2.0 * S))) - 1.0).

simpson(F, N, A, B, H) ->
    Inner_somme1 = fun(I) -> F(A + H * I) end,
    Inner_somme2 = fun(I) -> F(A + (I * H) + (H / 2.0)) end,
    S1 = somme(1, N, Inner_somme1),
    S2 = somme(0, N, Inner_somme2),

    abs((((B - A) / (6.0 * N)) * (F(A) + F(B) + (2.0 * S1 + 4.0 *
    S2))) - 1.0).

f1(_) -> 1.0.

f2(X) -> 2.0 * X.

f3(X) -> 3.0 * math:pow(X, 2.0).

f4(X) -> 4.0 * math:pow(X, 3.0).

f5(X) -> 5.0 * math:pow(X, 4.0).

f6(X) -> 6.0 * math:pow(X, 5.0).

f7(X) -> 1.0 / (2.0 * (math:sqrt(2.0) - 1.0) * math:sqrt(X + 1.0)).

f8(X) -> 2.0 / math:pow((X + 1.0), 2.0).

f9(X) -> (math:pi() / 2.0) * math:sin((math:pi() / 2.0) * X).

f10(X) -> math:exp(X) / (math:exp(1.0) - 1.0).

File: ruby/calcul.rb
module Calcul
    def self.calcul(n)
        return "n = #{n}\tReclangles\tTrapezes\tSimpson\n" +
            "f1\t" + calcul_integ(:f1, n) +
            "f2\t" + calcul_integ(:f2, n) +
            "f3\t" + calcul_integ(:f3, n) +
            "f4\t" + calcul_integ(:f4, n) +
            "f5\t" + calcul_integ(:f5, n) +
            "f6\t" + calcul_integ(:f6, n) +
            "f7\t" + calcul_integ(:f7, n) +
            "f8\t" + calcul_integ(:f8, n) +
            "f9\t" + calcul_integ(:f9, n) +
            "f10\t" + calcul_integ(:f10, n)
    end

    private
    def self.calcul_integ(f, n)
        a = 0.0
        b = 1.0
        h = (b - a) / n

        return sprintf("%.1E", rectangles(f, n, a, h)) + "\t\t" +
            sprintf("%.1E", trapezes(f, n, a, b, h)) + "\t\t" +
            sprintf("%.1E", simpson(f, n, a, b, h)) + "\n"
    end

    def self.rectangles(f, n, a, h)
        s = 0.0

        (0..n - 1).each do |i|
            s += method(f).call(a + h * i)
        end

        return ((h * s) - 1).abs
    end

    def self.trapezes(f, n, a, b, h)
        s = 0

        (1..n - 1).each do |i|
            s += method(f).call(a + (i * h))
        end

        return ((((b - a) / (2.0 * n)) * (method(f).call(a) + method(f).call(b) + (2.0 * s))) - 1).abs
    end

    def self.simpson(f, n, a, b, h)
        s1 = 0
        s2 = 0

        (1..n - 1).each do |i|
            s1 += method(f).call(a + (i * h))
        end

        (0..n - 1).each do |i|
            s2 += method(f).call(a + (i * h) + (h / 2.0))
        end

        return ((((b - a) / (6.0 * n)) * (method(f).call(a) + method(f).call(b) + (2.0 * s1 + 4.0 * s2))) - 1).abs
    end

    def self.f1(x)
        return 1
    end

    def self.f2(x)
        return 2.0 * x
    end

    def self.f3(x)
        return 3.0 * (x ** 2.0)
    end

    def self.f4(x)
        return 4.0 * (x ** 3.0)
    end

    def self.f5(x)
        return 5.0 * (x ** 4.0)
    end

    def self.f6(x)
        return 6.0 * (x ** 5.0)
    end

    def self.f7(x)
        return 1.0 / (2.0 * (Math.sqrt(2) - 1.0) * Math.sqrt(x + 1.0))
    end

    def self.f8(x)
        return 2.0 / ((x + 1.0) ** 2.0)
    end

    def self.f9(x)
        return (Math::PI / 2.0) * Math.sin((Math::PI / 2.0) * x)
    end

    def self.f10(x)
        return (Math::E ** x) / (Math::E - 1.0)
    end
end
